### PR TITLE
Add putwillblock() and takewillblock() to channels.jl

### PR DESCRIPTION
**Pull request by SamuraiAku**
This is a proposed resolution to issue #22863. Instead of creating an isfull() function for channels, which was not recommended, I created willblockput(). I also created the complementary function willblocktake().

### CODE REVIEW

File: base/channels.jl [hunk 1]
```diff
@@ -421,6 +421,20 @@ isready(c::Channel) = n_avail(c) > 0
 n_avail(c::Channel) = isbuffered(c) ? length(c.data) : length(c.cond_put.waitq)
 isempty(c::Channel) = isbuffered(c) ? isempty(c.data) : isempty(c.cond_put.waitq)
 
+"""
+    willblockput(c::Channel)
+
+Returns `true` if a [`put!`](@ref) on a [`Channel`](@ref) will block
+"""
+willblockput(c::Channel) =  isbuffered(c) ? (n_avail(c) == c.sz_max) : (length(c.cond_take.waitq) == 0)
```
Comment (jonas-schulze) [inline base/channels.jl:429]:
```suggestion
willblockput(c::Channel) =  isbuffered(c) ? (n_avail(c) == c.sz_max) : (length(c.cond_take.waitq) != 0)
```

but I might be mistaken. This should be tested.

**Review by jonas-schulze — COMMENTED** [review]
1. To me, the names rather suggest that they perform a put/take in a blocking way .. which is what `put`/`take!` do already. Maybe `putwillblock`/`takewillblock` are a better fit, especially in combination with an `if ...`?
2. I think `isfull` is a much more concise name for this, as mentioned in that issue. Why was it rejected? `full` or `Base.full` does not (no longer?) exist.

File: base/channels.jl (hunk 1 above)
Comment (SamuraiAku) [reply base/channels.jl:429]: 1. Changing to putwillblock is a good suggestion. 

2. isfull() is used for a circular buffer in DataStructues.jl and I think that’s where the feeling that a different name should come from. I can’t say I have a strong opinion on this one way or the other.

**Comment by SamuraiAku** [comment]
I've renamed the functions as suggested and added tests for buffered and unbuffered chanels.

**Comment by jonas-schulze** [comment]
> isfull() is used for a circular buffer in DataStructues.jl and I think that’s where the feeling that a different name should come from. I can’t say I have a strong opinion on this one way or the other.

I see. Having thought about it a bit more, I like your suggested names more now. An unbuffered channel could always be considered "full", even though a put wouldn't block.

**Comment by jonas-schulze** [comment]
What do you think about defining methods for `RemoteChannel`s? One could argue that networking is always blocking, or that the answer might no longer be valid by the time it arrives on the asking machine, but I like your heuristics ("either there is space or someone is waiting") and I think it would be a plus if `Channel`s and `RemoteChannel`s could be swapped easily. So I see four options:

* 👎 don't define them for `RemoteChannel`s
* 🚀 make them `...willblock(rr::RemoteChannel) = r.where == myid() && <your heuristic>`
* 🎉 make them `...willblock(rr::RemoteChannel) = <your heuristic>`
* 😕 make them `...willblock(rr::RemoteChannel) = true`

Either way, I'd say that should be a separate PR.

**Comment by SamuraiAku** [comment]
@jonas-schulze does the “need news” tag mean you want me to add something to NEWS.md? Or does someone else do that? If I do add something what section would this go under?

**Comment by SamuraiAku** [comment]
I can look into *willblock for remote channels. Since RemoteChannel already implements isready() then there is already a potential race condition and we’re no worse off than before.  If we do it there then we should do the same thing for Future as well. Maybe for WorkerPool and CachingPool in another PR for consistency since they all implement the same functions.

**Comment by SamuraiAku** [comment]
Added details to NEWS.md and the manual

**Review by jonas-schulze — COMMENTED** [review]
> @jonas-schulze does the “need news” tag mean you want me to add something to NEWS.md? Or does someone else do that? If I do add something what section would this go under?

Oh, I just read the "first time contributor" batch. Thanks for your contribution and welcome to the community! I should have given a few more explanations. 🙈

New features (like newly exported functions from Julia Base) need to be "announced" in NEWS.md, which is usually done by the contributor of the PR / new feature. The content of NEWS.md will become the changelog of the next release.

> Added details [...] the manual

I think it would be better to explain the new functions in the corresponding bullet points of `put!` and `take!` that already talk about blocking (instead of adding two new bullet points):

```
  * If a [`Channel`](@ref) is empty, readers (on a [`take!`](@ref) call) will block until data is available.
    This can be tested (without blocking) using [`takewillblock`](@ref).
  * If a [`Channel`](@ref) is full, writers (on a [`put!`](@ref) call) will block until space becomes available.
    This can be tested (without blocking) using [`putwillblock`](@ref).
```

Then, you need to also add the new functions to [doc/src/base/parallel.md](https://github.com/JuliaLang/julia/blob/master/doc/src/base/parallel.md) such that their docstrings are included in the manual. This is the reason CI fails currently.

> I can look into *willblock for remote channels. Since RemoteChannel already implements isready() then there is already a potential race condition and we’re no worse off than before. [...]

👍

> [...] If we do it there then we should do the same thing for Future as well. Maybe for WorkerPool and CachingPool in another PR for consistency since they all implement the same functions.

[Future](https://docs.julialang.org/en/v1/manual/distributed-computing/)s are used differently: they can only be assigned once, i.e. the first `put!` won't block (ignoring networking) and a second `put!` will error, so `putwillblock` isn't needed or doesn't really make sense. Then, `takewillblock` is not worth adding in my opinion (there shouldn't be too many ways to do the same thing).

WorkerPools and CachingPools ([AbstractWorkerPool](https://docs.julialang.org/en/v1/stdlib/Distributed/#Distributed.AbstractWorkerPool)s) are yet another thing. I don't have that much experience with them, but when I need a worker (another Julia process to do some heavy computations) (`take!`) or I want to make it available again (`put!`), I am not interested in whether those calls would block. I'd say `*willblock` would just encourage unnecessarily complex control flow.

**Comment by SamuraiAku** [comment]
Manual updated. Hopefully CI passes now.

On a related topic, once putwillblock and takewillblock are implemented on RemoteChannel is there any reason to keep isready()? The function takewillblock() provides the same information (it is an inverted alias of isready) with a more descriptive name. It should perhaps be considered for deprecation.

**Comment by jonas-schulze** [comment]
LGTM, thanks for all the updates. 🙂 

> Manual updated. Hopefully CI passes now.
> 
> On a related topic, once putwillblock and takewillblock are implemented on RemoteChannel is there any reason to keep isready()? The function takewillblock() provides the same information (it is an inverted alias of isready) with a more descriptive name. It should perhaps be considered for deprecation.

That won't happen, at least not in a v1.x release (and v2 isn't on the horizon yet). Furthermore, `takewillblock` is quite a technical name that -- I suppose -- won't be used all that often in user code. `isready` should be the more natural entry point there.

**Comment by SamuraiAku** [comment]
@jonas-schulze FYI I've created PR #40921 for adding putwillblock() and takewillblock() to a RemoteChannel.

**Comment by SamuraiAku** [comment]
Rebased to make it easier to follow this PR in conjunction with PR #40921

**Comment by JeffBezanson** [comment]
These functions seem like a bad idea to me --- don't they just encourage writing race conditions?

**Comment by SamuraiAku** [comment]
If a program was using both putwillblock() and takewillblock() on a channel then yes, there could very well be nasty race conditions. But the use of only one of them can be useful to programs. The choice of which one to use depends on the particular behavior of the threads in that program. Here are two possibilities:
1) A producer thread must be constantly servicing a hardware sensor at a relatively high rate to prevent buffer overflows, while the consumer thread is performing intensive calculations and cannot process all of the sensor data. In this situation the consumer thread blocks on take!() on an unbuffered channel when it is ready for a new data set. The producer thread tests if the consumer wants new data by invoking putwillblock().  If that test is false, the producer put!()'s the data into the channel without blocking and can immediately get back to servicing the hardware. If the test is true, the producer drops the data on the floor and goes back to servicing the hardware.
2) The consumer thread is pulling data from many producer threads each with its own channel, with each producer generating data at unpredictable intervals. The consumer does not want to block on a channel. It would use takewillblock() to test if there is data available. If that test is false, the consumer take!()'s the data immediately and moves on. If the test is true, the consumer passes over the channel and moves on.
 Note:  You could also do this with isready(), since takewillblock() is an inverted alias of isready(). But I think that the name is confusing since it's not clear from the name what "ready" means.  Is the channel ready to accept data (incorrect) or it is ready to produce data (correct)? This function name removes that ambiguity.

**Comment by touchft** [comment]
How about to extend "isready" to `put!`  while keep default to 'take' ?

For example:
isready(c::Channel, dir::Symbol=:take)
or
isready(c::Channel, dir::String="take")
so,
`isready(c, :put)` or `isready(c, "put")` is clear to ready and understant.

**Comment by tkf** [comment]
> 1\. A producer thread must be constantly servicing a hardware sensor at a relatively high rate to prevent buffer overflows,

Channel is about synchronization. Introducing synchronization inside latency-sensitive part of your code does not seem a good idea. I'd recommend using a nonblocking (possibly bounded/leaky) queue.

> 2\. The consumer does not want to block on a channel. It would use takewillblock() to test if there is data available. If that test is false, the consumer take!()'s the data immediately and moves on.

There is no guarantee that `take!` succeeds since another task can be doing the same thing. You'd need `maybetake!(::Channel{T}) -> result::Union{Nothing,Some{T}}` here.

I'm sympathetic to the motivation behind this kind of "racy" construct. Concurrency is about controlling the timing and it can be usable and useful sometimes. But the proposed API does not seem a particularly composable one that promotes correct code. For example, `takewillblock`+`take!` is not `maybetake!`. Julia lacks advanced synchronization protocols. While I agree we need more of those, I'd prefer seeing cohesive design behind the protocols. Personally, I'm interested in importing ideas from Reagents ([paper PDF](https://aturon.github.io/academic/reagents.pdf)) and [Concurrent ML](https://en.wikipedia.org/wiki/Concurrent_ML) for something like Go's `select` but more composable and extensible.

**Comment by SamuraiAku** [comment]
> > 1. A producer thread must be constantly servicing a hardware sensor at a relatively high rate to prevent buffer overflows,
> 
> Channel is about synchronization. Introducing synchronization inside latency-sensitive part of your code does not seem a good idea. I'd recommend using a nonblocking (possibly bounded/leaky) queue.

@tkf Where would I find a thread-safe non-blocking leaky queue? That would solve this issue, but I'm not seeing anything like that in the documentation.

> > 2. The consumer does not want to block on a channel. It would use takewillblock() to test if there is data available. If that test is false, the consumer take!()'s the data immediately and moves on.
> 
> There is no guarantee that `take!` succeeds since another task can be doing the same thing. You'd need `maybetake!(::Channel{T}) -> result::Union{Nothing,Some{T}}` here.
> 

The scenario I'm envisioning has a single consumer thread, so if false == takewillblock() then you can grab the data without worry.

> I'm sympathetic to the motivation behind this kind of "racy" construct. Concurrency is about controlling the timing and it can be usable and useful sometimes. But the proposed API does not seem a particularly composable one that promotes correct code. For example, `takewillblock`+`take!` is not `maybetake!`. Julia lacks advanced synchronization protocols. While I agree we need more of those, I'd prefer seeing cohesive design behind the protocols. Personally, I'm interested in importing ideas from Reagents ([paper PDF](https://aturon.github.io/academic/reagents.pdf)) and [Concurrent ML](https://en.wikipedia.org/wiki/Concurrent_ML) for something like Go's `select` but more composable and extensible.

Yeah I don't doubt this could be improved over the long-term. 

But is there something small that we could do now that would be of benefit to users?  That was my motivation behind this PR.  It seemed like a logical enhancement to the functionality already provided by isready(). What @touchft proposed where isready() is extended to support both put! and take! is another alternative that is maybe better from a user perspective.

**Comment by KronosTheLate** [comment]
Based on https://github.com/JuliaLang/julia/issues/22863 I created https://github.com/JuliaLang/julia/pull/53159, which I believe should closed this PR if it goes through.

**Comment by SamuraiAku** [comment]
Closed in favor of #53159

**Pull request closed without merging**